### DISCUSSION TITLE
[Brigade] Storage Tests

### DISCUSSION THREAD

**Discussion by nate-double-u** [cncf/mentoring#601] [Google Summer of Code 2022]
Brigade's data access packages are currently unit-tested against mock implementations of MongoDB client interfaces. These tests have been adequate for asserting that queries and statements are constructed properly (look like we think they should) and that mock query and statement results can be unmarshaled without error into domain types, but this approach cannot assert that DB queries and statements are logically correct and actually achieve the desired results, since a live database would be required to accomplish that. Given the importance of the data access code, we would like to develop a new suite of integration tests directly targeted at that code. These new tests should run against a live (and disposable) MongoDB database and assert that all queries and statements achieve the desired results. If the new suite of tests exposes bugs in the existing data access code, correcting those bugs is within the scope of this project as well.

Mentor(s): Kent Rancourt (@krancour)

https://github.com/cncf/mentoring/blob/main/summerofcode/2022.md#storage-tests

**Comment by krancour** [comment]
Hey @nate-double-u. I'm unclear on status / next steps. Were all the project ideas for which you opened a discussion accepted?

**Reply by nate-double-u** [reply]
Hi @krancour, no we're still in the discussion phase. Right now potential GSoC contributors should be discussing application ideas with the projects. That's what these discussions are intended to help with.

The next key date in the process is April 4 (at 18:00 UTC), when the GSoC contributor application period begins.

**Reply by krancour** [reply]
So refer people to these discussions when they inquire?

**Reply by nate-double-u** [reply]
These discussions, or your [upstream issue](https://github.com/brigadecore/brigade/issues/1811). I probably should have linked them more closely, but either is fine. 

Also, I had meant to link it in my previous message, but here's the GSoC timeline: https://developers.google.com/open-source/gsoc/timeline

**Comment by krancour** [comment]
Awesome. Thanks for the info @nate-double-u!

**Reply by nate-double-u** [reply]
Happy to help! 😃

**Comment by chebro** [comment]
hello! i have been in touch with @krancour in slack over the last three weeks and have created a draft proposal that i would like to get feedback on to improve it before making the final submission. 

@nate-double-u is there an official process to request reviews from the mentors?

**Comment by AnuragThePathak** [comment]
Hi, I had a confusion, do we need to replace all mock implementations by live database queries or need to deal with only the tests for the functions which execute database queries using the mongodb driver?

**Reply by krancour** [reply]
I'm not clear yet on whether we would opt to replace the existing tests or keep them, but either way, we're talking about a whole new test suite that tests the entirety of the data access layer against a live database.

**Reply by AnuragThePathak** [reply]
@krancour should we write one integration test per mongodb query? And how many test cases to include per integration test ideally?

**Reply by AnuragThePathak** [reply]
I had one more doubt don't we need to write integration tests for the v2/internal/api/mongodb package only?

**Reply by krancour** [reply]
It's only that package, yes, It's _at least_ on test case per query or statement.

**Reply by AnuragThePathak** [reply]
And for the live database, should we use the official mongodb docker image or bitnami mongodb docker image?

**Reply by AnuragThePathak** [reply]
Can the build tags for current suit integration tests be changed to end-to-end tests (e2e for example)? Or we should keep it as integration only? I was thinking of using different package and TestMain for new suite of integration tests.

**Reply by AnuragThePathak** [reply]
And do we need to test the NewXStore functions (like NewProjectsStore, NewEventsStore) ?

**Reply by krancour** [reply]
I think we're getting too far into implementation details here.

**Reply by AnuragThePathak** [reply]
Oh I thought such details are required 😅.

**Reply by krancour** [reply]
Myself and another maintainer will have some say in what applications are accepted, but the people with final say, as I understand it, are CNCF org-level folks who do not necessarily have any in-depth familiarity with Brigade. Detail is good, but if we're down to the level of talking about file names, tags, specific functions, etc., those details won't mean anything to the people making the final call.

**Reply by AnuragThePathak** [reply]
Oh, while writing my 2nd proposal which is for storage tests, I was trying to explain my strategy on almost all functions for which we need to write tests. Now after knowing this I'll try to avoid that as far as possible.

**Reply by chebro** [reply]
@krancour i have updated the timeline and uploaded the same to the gsoc website. i have sent a request on slack for a review, please let me know what your thoughts are.

**Comment by AnuragThePathak** [comment]
Will it be a good idea to add the table of contents to the proposal pdfs?

**Reply by nate-double-u** [reply]
I'm sorry I missed this before the submission deadline. Hopefully you submitted what you thought would be best.

Generally on short or well structured documents I don't tend to include a table of contents, it depends on how complicated it is.

**Reply by AnuragThePathak** [reply]
> Hopefully you submitted what you thought would be best.

Yeah I thought that having that will help reviewers so I added that.

> I'm sorry I missed this before the submission deadline.

No issues. Actually, it's my fault that I didn't clear doubts beforehand.
